pkg/ring/client/pool.go

```diff
@@ -0,0 +1,198 @@
+	return client, ok
+}
+
+// GetClientFor gets the client for the specified address. If it does not exist it will make a new client
+// at that address
+func (p *Pool) GetClientFor(addr string) (PoolClient, error) {
+	client, ok := p.fromCache(addr)
+	if ok {
+		return client, nil
+	}
+
+	p.Lock()
+	defer p.Unlock()
+	client, ok = p.clients[addr]
+	if ok {
+		return client, nil
+	}
+
+	client, err := p.factory(addr)
+	if err != nil {
+		return nil, err
+	}
+	p.clients[addr] = client
+	if p.clientsMetric != nil {
+		p.clientsMetric.Add(1)
+	}
+	return client, nil
+}
+
+// RemoveClientFor removes the client with the specified address
+func (p *Pool) RemoveClientFor(addr string) {
+	p.Lock()
+	defer p.Unlock()
+	client, ok := p.clients[addr]
+	if ok {
+		delete(p.clients, addr)
+		if p.clientsMetric != nil {
+			p.clientsMetric.Add(-1)
+		}
+		// Close in the background since this operation may take awhile and we have a mutex
+		go func(addr string, closer PoolClient) {
+			if err := closer.Close(); err != nil {
+				level.Error(p.logger).Log("msg", fmt.Sprintf("error closing connection to %s", p.serviceName), "addr", addr, "err", err)
+			}
+		}(addr, client)
+	}
+}
+
+// RegisteredAddresses returns all the addresses that a client is cached for
+func (p *Pool) RegisteredAddresses() []string {
+	result := []string{}
+	p.RLock()
+	defer p.RUnlock()
+	for addr := range p.clients {
+		result = append(result, addr)
+	}
+	return result
+}
+
+// Count returns how many clients are in the cache
+func (p *Pool) Count() int {
+	p.RLock()
+	defer p.RUnlock()
+	return len(p.clients)
+}
+
+func (p *Pool) removeStaleClients() {
+	clients := map[string]struct{}{}
+	replicationSet, err := p.ring.GetAll()
+	if err != nil {
+		level.Error(util.Logger).Log("msg", "error removing stale clients", "err", err)
+		return
+	}
+
+	for _, ing := range replicationSet.Ingesters {
+		clients[ing.Addr] = struct{}{}
+	}
+
+	for _, addr := range p.RegisteredAddresses() {
+		if _, ok := clients[addr]; ok {
+			continue
+		}
+		level.Info(util.Logger).Log("msg", "removing stale client", "addr", addr)
+		p.RemoveClientFor(addr)
+	}
+}
+
+// cleanUnhealthy loops through all servers and deletes any that fails a healthcheck.
+func (p *Pool) cleanUnhealthy() {
+	for _, addr := range p.RegisteredAddresses() {
+		client, ok := p.fromCache(addr)
+		// not ok means someone removed a client between the start of this loop and now
+		if ok {
+			err := healthCheck(client, p.cfg.HealthCheckTimeout)
+			if err != nil {
+				level.Warn(util.Logger).Log("msg", fmt.Sprintf("removing %s failing healthcheck", p.serviceName), "addr", addr, "reason", err)
+				p.RemoveClientFor(addr)
+			}
+		}
+	}
+}
+
+// healthCheck will check if the client is still healthy, returning an error if it is not
+func healthCheck(client PoolClient, timeout time.Duration) error {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	ctx = user.InjectOrgID(ctx, "0")
```

Comment:
I wonder why we need to do this (unrelated to this PR).

Reply:
It's required by the `ServerUserHeaderInterceptor` middleware for any gRPC call.